src/theory/datatypes/infer_proof_cons.cpp

```diff
@@ -286,16 +305,50 @@ void InferProofCons::convert(InferenceId infer, TNode conc, TNode exp, CDProof*
         premises.push_back(sconc);
         std::vector<Node> pols;
         std::vector<Node> lits;
-        for (const Node& e : expv)
+        std::map<Node, Node>::iterator itt;
+        for (const Node& e : sconc)
         {
-          if (e.getKind() != Kind::NOT || e[0].getKind() != Kind::APPLY_TESTER)
+          if (e == conc)
+          {
+            continue;
+          }
+          Node en = e.notNode();
+          premises.push_back(en);
+          pols.emplace_back(truen);
+          lits.emplace_back(e);
+          // must ensure we have a proof of en
+          Assert(e.getKind() == Kind::APPLY_TESTER);
+          bool successLit = false;
+          itt = tmap.find(e.getOperator());
+          if (itt != tmap.end())
+          {
+            if (itt->second == en)
+            {
+              successLit = true;
+            }
+            else
+            {
+              // otherwise maybe provable modulo equality?
+              // This is to handle e.g.
+              // (and (not (is-cons x)) (= x y)) => (is-nil y)
```

Comment:
For this all to be more clear it'd be best to already do, in the first example, the most general form of what this elaboration is doing. It is a bit hard to do the connection only when you get here.

Reply:
Added more doc